lib/ts/recipe/session/sessionAuth.tsx

```diff
@@ -30,142 +32,127 @@ type PropsWithoutAuth = {
 
 type PropsWithAuth = {
     requireAuth: true;
-    redirectToLogin: () => void;
+    redirectToLogin?: () => void;
 };
 
 type Props = (PropsWithoutAuth | PropsWithAuth) & {
     onSessionExpired?: () => void;
 };
 
+const noop = () => {
+    return;
+};
+
 const SessionAuth: React.FC<Props> = ({ children, ...props }) => {
-    if (props.requireAuth === true && props.redirectToLogin === undefined) {
-        throw new Error("You have to provide redirectToLogin function when requireAuth is true");
+    if (props.requireAuth === true) {
+        if (props.redirectToLogin === undefined && props.onSessionExpired === undefined) {
+            throw new Error(
+                "You have to provide redirectToLogin or onSessionExpired function when requireAuth is true"
+            );
+        }
     }
 
     const parentSessionContext = useContext(SessionContext);
     const [context, setContext] = useState<SessionContextType | undefined>(undefined);
 
     const session = useRef(Session.getInstanceOrThrow());
 
-    const setInitialContext = useCallback(async (): Promise<void> => {
-        const sessionExists = await doesSessionExist();
+    const redirectToLogin = props.requireAuth
+        ? props.redirectToLogin !== undefined
+            ? props.redirectToLogin
+            : noop
+        : noop;
 
-        if (sessionExists === false) {
-            return setContext({
-                doesSessionExist: sessionExists,
-                jwtPayload: {},
-                userId: "",
-            });
-        }
+    // on mount
+    useEffect(() => {
+        let cancelUseEffect = false;
 
-        const [jwtPayload, userId] = await Promise.all([getJWTPayloadSecurely(), getUserId()]);
+        const buildContext = async (): Promise<SessionContextType> => {
+            if (hasParentProvider(parentSessionContext)) {
+                return parentSessionContext;
+            }
+
+            const sessionExists = await doesSessionExist();
 
-        return setContext({
-            doesSessionExist: sessionExists,
-            jwtPayload,
-            userId,
-        });
+            if (sessionExists === false) {
+                return {
+                    doesSessionExist: false,
+                    jwtPayload: {},
+                    userId: "",
+                };
+            }
+
+            return {
+                doesSessionExist: true,
+                jwtPayload: await getJWTPayloadSecurely(),
+                userId: await getUserId(),
+            };
+        };
+
+        async function setInitialContextAndMaybeRedirect() {
+            // if this component is unmounting, or the context has already
+            // been set, then we don't need to proceed...
+            if (cancelUseEffect || context !== undefined) {
```

Comment:
This check is supposed to happen AFTER the `await` cause using the `await` can add a race condition..